File: 1055-shortest-way-to-form-string/1055-shortest-way-to-form-string.cpp
class Solution {
public:
    int shortestWay(string source, string target) {
        unordered_map<char,int> hashMap;
        for(char c : source) hashMap[c]++; 
        int start = 0;
        int end = 0; 
        int answer = 0; 
        while(end < target.length()){
            if(!hashMap.count(target[end])) return -1; 

            while(start < source.length() && end < target.length()){
                if(source[start] == target[end]){
                    start++;
                    end++; 
                } else{
                    start++; 
                }
            }

            answer++;
            if(start >= source.length()){
                start = 0; 
            }
            
        }
        return answer;
    }
};

File: 1055-shortest-way-to-form-string/1055-shortest-way-to-form-string.java
class Solution {
    public int shortestWay(String source, String target) {
        Set<Character> hashSet = new HashSet<>(); 
        for(char c : source.toCharArray()) hashSet.add(c); 

        int count = 0; 
        int start = 0, end = 0; 
        while(end < target.length()){
            if(!hashSet.contains(target.charAt(end))) return -1; 
            int startIndex = end; 
            while(start < source.length() && end < target.length()){
                if(source.charAt(start) == target.charAt(end)){
                    end++; 
                }
                start++; 
            }

            if(end == startIndex){
                return -1; 
            }

            start = 0; 
            count++; 
        }

        return count; 
    }
}